File: logs_analyzer/app/java8/src/main/java/com/databricks/apps/logs/LogAnalyzerTotal.java
package com.databricks.apps.logs;

import java.io.Serializable;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

import scala.Tuple2;
import scala.Tuple4;

import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;

public class LogAnalyzerTotal implements Serializable {
  // These static variables stores the running content size values.
  private final AtomicLong runningCount = new AtomicLong(0);
  private final AtomicLong runningSum = new AtomicLong(0);
  private final AtomicLong runningMin = new AtomicLong(Long.MAX_VALUE);
  private final AtomicLong runningMax = new AtomicLong(Long.MIN_VALUE);
  private List<Tuple2<Integer, Long>> currentResponseCodeCounts = Collections.emptyList();
  private List<String> currentIPAddresses = Collections.emptyList();
  private List<Tuple2<String, Long>> currentTopEndpoints = Collections.emptyList();

  public void processAccessLogs(JavaDStream<ApacheAccessLog> accessLogsDStream) {
    // Calculate statistics based on the content size, and update the static variables to track this.
    accessLogsDStream.foreachRDD(accessLogs -> {
          Tuple4<Long, Long, Long, Long> stats =
              Functions.contentSizeStats(accessLogs);
          if (stats != null) {
            runningCount.getAndAdd(stats._1());
            runningSum.getAndAdd(stats._2());
            runningMin.set(Math.min(runningMin.get(), stats._3()));
            runningMax.set(Math.max(runningMax.get(), stats._4()));
          }
        }
    );

    // A DStream of Response Code Counts;
    JavaPairDStream<Integer, Long> responseCodeCountDStream = accessLogsDStream
        .transformToPair(Functions::responseCodeCount)
        .updateStateByKey(Functions.COMPUTE_RUNNING_SUM);
    responseCodeCountDStream.foreachRDD(rdd -> currentResponseCodeCounts = rdd.take(100));

    // A DStream of ipAddressCounts.
    JavaDStream<String> ipAddressesDStream = accessLogsDStream
        .transformToPair(Functions::ipAddressCount)
        .updateStateByKey(Functions.COMPUTE_RUNNING_SUM)
        .transform(Functions::filterIPAddress);
    ipAddressesDStream.foreachRDD(rdd -> currentIPAddresses = rdd.take(100));

    // A DStream of endpoint to count.
    JavaPairDStream<String, Long> endpointCountsDStream = accessLogsDStream
        .transformToPair(Functions::endpointCount)
        .updateStateByKey(Functions.COMPUTE_RUNNING_SUM);
    endpointCountsDStream.foreachRDD(rdd -> currentTopEndpoints = rdd.top(10,
        new Functions.ValueComparator<>(Comparator.<Long>naturalOrder())));
  }

  public LogStatistics getLogStatistics() {
    if (runningCount.get() == 0) {
      return LogStatistics.EMPTY_LOG_STATISTICS;
    }

    return new LogStatistics(new Tuple4<>(runningCount.get(), runningSum.get(),
        runningMin.get(), runningMax.get()),
        currentResponseCodeCounts, currentIPAddresses, currentTopEndpoints);
  }
}

File: logs_analyzer/app/scala/src/main/scala/com/databricks/apps/logs/LogAnalyzerTotal.scala
package com.databricks.apps.logs

import java.util.concurrent.atomic.AtomicLong

import scala.math._

import org.apache.spark.streaming.dstream.DStream

/** Collects total statistics from the moment application started. */
class LogAnalyzerTotal extends AnalyzeFunctions with Serializable {
  val runningCount = new AtomicLong(0)
  val runningSum = new AtomicLong(0)
  val runningMin = new AtomicLong(0)
  val runningMax = new AtomicLong(0)
  var currentResponseCodeCounts: Map[Int, Long] = Map.empty
  var currentIPAddresses: Array[String] = Array.empty
  var currentTopEndpoints: Map[String, Long] = Map.empty

  def processAccessLogs(accessLogsDStream: DStream[ApacheAccessLog]): Unit = {
    // Calculate statistics based on the content size, and update the static variables to track this.
    accessLogsDStream.foreachRDD(accessLogs => {
      val stats = contentSizeStats(accessLogs)
      stats.foreach(s => {
        runningCount.getAndAdd(s._1)
        runningSum.getAndAdd(s._2)
        runningMin.set(min(runningMin.get(), s._3))
        runningMax.set(max(runningMax.get(), s._4))
      })
    })

    // A DStream of Response Code Counts;
    val responseCodeCountDStream: DStream[(Int, Long)] = accessLogsDStream
      .transform(responseCodeCount)
      .updateStateByKey(computeRunningSum)
    responseCodeCountDStream.foreachRDD(rdd => currentResponseCodeCounts = rdd.take(100).toMap)

    // A DStream of ipAddressCounts.
    val ipAddressesDStream: DStream[String] = accessLogsDStream
      .transform(ipAddressCount)
      .updateStateByKey(computeRunningSum)
      .transform(filterIPAddress)
    ipAddressesDStream.foreachRDD(rdd => currentIPAddresses = rdd.take(100))

    // A DStream of endpoint to count.
    val endpointCountsDStream: DStream[(String, Long)] = accessLogsDStream
      .transform(endpointCount)
      .updateStateByKey(computeRunningSum)
    endpointCountsDStream
      .foreachRDD(rdd => currentTopEndpoints = rdd.top(10)(Ordering.by[(String, Long), Long](_._2)).toMap)
  }

  def getLogStatistics: LogStatistics = {
    LogStatistics(
      (runningCount.get(),
        runningSum.get(),
        runningMin.get(),
        runningMax.get()),
      currentResponseCodeCounts,
      currentIPAddresses,
      currentTopEndpoints)
  }
}
